tensorboard/plugins/histogram/summary.py

```diff
@@ -0,0 +1,201 @@
+        right_edges = edges[1:]
+        return tf.transpose(tf.stack(
+          [left_edges, right_edges, bucket_counts]))
+
+      def when_singular():
+        center = min_
+        bucket_starts = tf.stack([center - 0.5])
+        bucket_ends = tf.stack([center + 0.5])
+        bucket_counts = tf.stack([tf.cast(tf.size(data), tf.float64)])
+        return tf.transpose(
+          tf.stack([bucket_starts, bucket_ends, bucket_counts]))
+
+      return tf.cond(is_singular, when_singular, when_nonsingular)
+
+    return tf.cond(is_empty, when_empty, when_nonempty)
+
+
+def op(name,
+       data,
+       bucket_count=None,
+       display_name=None,
+       description=None,
+       collections=None):
+  """Create a histogram summary op.
+
+  Arguments:
+    name: A unique name for the generated summary node.
+    data: A `Tensor` of any shape. Must be compatible with `float64`.
+    bucket_count: Optional positive `int` or scalar `int32` `Tensor`.
+    display_name: Optional name for this summary in TensorBoard.
+      Defaults to `name`.
+    description: Optional long-form description for this summary.
+      Markdown is supported.
+    collections: Optional list of graph collections keys. The new
+      summary op is added to these collections. Defaults to
+      `[Graph Keys.SUMMARIES]`.
+
+  Returns:
+    A TensorFlow summary op.
+  """
+  if display_name is None:
+    display_name = name
+  summary_metadata = metadata.create_summary_metadata(
+    display_name=display_name, description=description)
+  with tf.name_scope(name):
+    tensor = _buckets(data, bucket_count=bucket_count)
+    return tf.summary.tensor_summary(name='histogram_summary',
+                                     tensor=tensor,
+                                     collections=collections,
+                                     summary_metadata=summary_metadata)
+
+
+def pb(name, data, bucket_count=None, display_name=None, description=None):
+  """Create a histogram summary protobuf.
+
+  Arguments:
+    name: A unique name for the generated summary, including any desired
+      name scopes.
+    data: A `np.array` or array-like form of any shape. Must have type
+      compatible with `float`.
+    bucket_count: Optional positive `int`.
+    display_name: Optional name for this summary in TensorBoard.
+      Defaults to `name`.
+    description: Optional long-form description for this summary.
+      Markdown is supported.
+  """
+  if bucket_count is None:
+    bucket_count = BUCKETS
+  if bucket_count <= 0:
+    raise ValueError('assertion failed: "bucket_count" must be positive, '
+                     'but was %s' % bucket_count)
+  data = np.array(data).flatten().astype(float)
+  if data.size == 0:
+    buckets = np.array([]).reshape((0, 3))
+  else:
+    min_ = np.min(data)
+    max_ = np.max(data)
+    range_ = max_ - min_
+    if range_ == 0:
+      center = min_
+      buckets = np.array([[center - 0.5, center + 0.5, float(data.size)]])
```

Comment:
A couple mitigating points: (a) it's [well documented](https://github.com/tensorflow/tensorboard/pull/246/files#diff-a5b9ad5bed9db3883123a023e79faa93R21-25); (b) the realistic cases where `range == 0` or especially `data.size == 0` are surely very rare; and (c) the end-to-end effect is the same (i.e., if all you care about is "what does the `vz-histogram-chart` display as a result of my TensorFlow code?" then you'll see no difference).

If you're computing a histogram in the first place, I think you'll be more surprised about why your data is all zeros or absent than the edge cases of the bucketing mechanism.

Adding empty buckets is possible, though I don't like the fact that we'd have to add 14 buckets on one side and 15 buckets on the other, or enforce a parity constraint. (I considered this, and the parity constraint is why I didn't do it in the first place.)

I think that returning an empty set of buckets when there is no data is the most reasonable thing to do: in the singleton case, we at least have an anchor point for the datum (though the "unit size" parameter is arbitrary[1]), but creating, say, a bunch of unit-size buckets around, say, zero doesn't sound good. (Suppose that my histogram data, when it exists, is always a small probability, and so I naturally try to take the logs of all of my bucket boundaries to plot them…but now I have no data so my buckets are like `[-3, -2]`…)

[1]: You may note that the problem that I describe re: logarithms could also happen here, so a better transformation would be `x => {left: x / 2, right: x * 3 / 2}` with appropriate handling for negative values, but then we still don't have a zero case.)

> How do you feel about making 29 empty buckets of unit size?

I'm inclined to make no change, but could be convinced otherwise if you feel strongly about this.

> If I give a bucket count, I'll be confused if that parameter is not respected.